hardware/detect.py

```diff
@@ -751,44 +832,35 @@ def get_cpus(hw_lst):
             processor), 'flags', lscpu['Flags']))
 
     hw_lst.append(('cpu', 'logical', 'number', int(lscpu['CPU(s)'])))
-    # Govenors could be differents on logical cpus
-    for cpu in range(int(lscpu['CPU(s)'])):
-        scaling_governor = "/sys/devices/system/cpu/cpufreq/policy{}/scaling_governor".format(
-            cpu)
-        if os.path.exists(scaling_governor):
-            with open(scaling_governor) as governorfile:
-                hw_lst.append(('cpu', "logical_{}".format(cpu),
-                               "governor", governorfile.readline().rstrip('\n')))
 
-    # Extracting numa nodes
-    hw_lst.append(('numa', 'nodes', 'count', int(lscpu['NUMA node(s)'])))
-    for numa in range(int(lscpu['NUMA node(s)'])):
-        cpus = lscpu['NUMA node{} CPU(s)'.format(numa)]
-        # lscpu -x provides the cpu mask
-        cpu_mask = lscpux['NUMA node{} CPU(s)'.format(numa)]
-        total_cpus = 0
-        min_cpu = None
-        max_cpu = None
+    _get_cpufreq_govenors(hw_lst, int(lscpu['CPU(s)']))
+    _get_cpu_numa_layout(hw_lst, lscpu, lscpux)
 
-        for item in cpus.split(','):
-            # lscpu report numa nodes like 0-5,48-53
-            if "-" in item:
-                max_cpu = int(item.split("-")[1])
-                min_cpu = int(item.split("-")[0])
-                total_cpus = total_cpus + max_cpu - min_cpu + 1
-            else:
-                # or like 0,1
-                if min_cpu is None:
-                    min_cpu = int(item)
-                else:
-                    max_cpu = int(item)
-                    total_cpus = total_cpus + max_cpu - min_cpu + 1
 
-        # total_cpus = 12 for "0-5,48-53"
-        hw_lst.append(('numa', 'node_{}'.format(
-            numa), 'cpu_count', total_cpus))
-        hw_lst.append(('numa', 'node_{}'.format(
-            numa), 'cpu_mask', cpu_mask))
+def get_cpus(hw_lst):
+    # Extracting lspcu information
+    lscpu = {}
+    output = detect_utils.output_lines('LANG=en_US.UTF-8 lscpu')
+
+    for line in output:
+        if ':' in line:
+            item, value = line.split(':')
+            lscpu[item.strip(':')] = value.strip()
+
+    # Extracting lspcu -x information
+    lscpux = {}
+    output = detect_utils.output_lines('LANG=en_US.UTF-8 lscpu -x')
+
+    for line in output:
+        if ':' in line:
+            item, value = line.split(':')
+            lscpux[item.strip(':')] = value.strip()
+
+    arch = lscpux.get('Architecture')
+    if arch == 'x86_64':
+        _x86_64_cpu_info(hw_lst, lscpu, lscpux)
+    elif arch == 'ppc64le':
+        _ppc64le_cpu_info(hw_lst, lscpu, lscpux)
 
```

Comment:
else? Should we maybe fall back to the x86_64 approach? Or at least do it for i686?

Reply:
I'm working on making things more generic so that it mor ort less works on both CPUs and 'should' mostly work on say aarch64